### ISSUE TITLE
Make pre-commit hook only look at changed files

### ISSUE THREAD

**Issue by dinomite** [jeremymailen/kotlinter-gradle#155]
kotlinter-gradle has most of the wiring necessary for a pre-commit hook, but [it was disabled](https://github.com/jeremymailen/kotlinter-gradle/pull/137/#pullrequestreview-427739206) because it checks _all_ files, not just those that are part of the to-be-committed changeset.

Make the pre-commit hook only examine files that are part of the changeset and wire it as a new task.

**Comment by jeremymailen** [comment]
Yeah, it's a good idea. Been meaning to get to this, but after all the work to get kotlin 1.4 compatibility out I took a bit of a hiatus.

**Comment by Noctiflore** [comment]
I saw you plan to do it using a git diff command to get the list of changed files, but modifying files and staging them without staging unwanted modifications from the working directory would still be hard.
Why don't stash the workdir (and unstash before real commit), as done [here](https://medium.com/@alistair.cerio/android-ktlint-and-pre-commit-git-hook-5dd606e230a9) ? git diff can still be used to check if the hook needs to be run at all, and if there are unstaged modifications to stash.

**Comment by jeremymailen** [comment]
Thank you @Noctiflore, these are good references. Will definitely try to make use of them.

**Comment by Noctiflore** [comment]
Some other useful readings and details:
- [my first link](https://medium.com/@alistair.cerio/android-ktlint-and-pre-commit-git-hook-5dd606e230a9) is a good starting point but fails with partially staged files.
- [git-format-staged](https://github.com/hallettj/git-format-staged) discusses the different alternatives ; a bit outdated, but still useful.
- [one of the answers here](https://stackoverflow.com/questions/8506334/what-to-stage-after-a-pre-commit-hook-was-run) propose to use git merge -Xtheirs instead of git stash pop, which may not be what is expected (overwrite linter changes)
- [this article](https://www.darrenlester.com/blog/git-pre-commit-stash) propose a good solution to handle a stash, and the specific case when stash push is a no-op.

I ended up writing this, inspired from these articles and git documentation:
```shell
#!/bin/bash

# Git pre-commit hook applying Ktlint on staged files
# (using user-defined gradle task)

STASH_NAME=pre-commit-$(date +%s)-$RANDOM
if ! git diff-files --quiet ; then
    # dirty workdir: stash to work on staged files only
    git stash push -q --keep-index -m "$STASH_NAME"
fi

echo "Pre-commit hook: Running ktlint (fix)"
gradle ktlintFix
RESULT=$?
# stage modified files
git add --update

# restore workdir
# get stash ID (stash{#}) (empty if workdir was clean)
STASH_ID=$(git stash list | sed -rn "s/(stash@\{.*\}).*$STASH_NAME.*/\1/p")

if [[ -n "$STASH_ID" ]] ; then
    # git stash pop would conflict with changes kept in index
    # in partially added files : use diff/apply.
    git diff $STASH_ID^2 $STASH_ID | git apply
    if [ $? -eq 0 ] ; then
        git stash drop -q $STASH_ID
    else
        echo "Pre-commit hook did conflicting changes (see previous messages)"
        echo "Previous workdir state is stashed in $STASH_ID: $STASH_NAME"
    fi
fi

exit $RESULT
```

**Comment by Mahoney** [comment]
For reference [ktlint's own pre-commit hook](https://github.com/pinterest/ktlint/blob/master/ktlint/src/main/resources/ktlint-git-pre-commit-hook.sh) is basically a one-liner:
```bash
git diff --name-only --cached --relative | grep '\.kt[s"]\?$' | xargs ktlint --relative .
```
Adding a `-F` to the ktlint command would add formatting. `git diff --cached` prevents unstaged files being included.

If the `formatKotlin` task took an `--includes` [command line flag](https://docs.gradle.org/current/userguide/custom_tasks.html#sec:declaring_and_using_command_line_options) of type `List<String>` that specified the specific files to include it could take the same approach?
```bash
echo "a b c" | xargs printf -- '-f %s\n' | xargs mycommand
git diff --name-only --cached --relative | grep '\.kt[s"]\?$' | xargs printf -- '--includes %s\n' | xargs $GRADLEW formatKotlin 
```

**Comment by Mahoney** [comment]
Actually it's a bit more complicated than that... I ended up with:
```bash
set -euo pipefail
changed=$(git diff --name-only --cached --relative | grep '\.kt[s"]\?$' || true)
echo "$changed" | xargs ktlint -F --relative .
echo "$changed" | xargs git add
if git diff --name-only --cached --quiet; then
  # After formatting, no files were changed! Exit, otherwise we get an empty commit
  exit 1
fi
```
Or in my imaginary future version of this plugin:
```bash
set -euo pipefail
changed=$(git diff --name-only --cached --relative | grep '\.kt[s"]\?$' || true)
echo "$changed" | xargs printf -- '--includes %s\n' | xargs $GRADLEW formatKotlin 
echo "$changed" | xargs git add
if git diff --name-only --cached --quiet; then
  # After formatting, no files were changed! Exit, otherwise we get an empty commit
  exit 1
fi
```

**Comment by jeremymailen** [comment]
Yeah, this is useful, thank you. Today the `formatKotlin` task is a parent of the individual source set format tasks, but we could implement a standalone `formatKotlinFiles` task accepting command args and otherwise unattached to the build lifecycle.

**Comment by KotlinIsland** [comment]
This is how ktlint-gradle does it
https://github.com/JLLeitschuh/ktlint-gradle/blob/master/plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/GitHook.kt
```kt
    diff=.git/unstaged-ktlint-git-hook.diff
    git diff --color=never > ${'$'}diff
    if [ -s ${'$'}diff ]; then
      git apply -R ${'$'}diff
    fi
    
    ${generateGradleCommand(taskName, gradleRootDirPrefix)}
    echo "Completed ktlint run."
    ${postCheck(shouldUpdateCommit)}
    
    if [ -s ${'$'}diff ]; then
      git apply --ignore-whitespace ${'$'}diff
    fi
    rm ${'$'}diff
    unset diff
```

**Comment by mvarnagiris** [comment]
Is this still something that is being considered to be done? Right now for me running `formatKotlin` or `lintKotlin` adds about 22 seconds each on my machine on my large project. So pre-push hook extends the whole push process by more than 40 seconds. And this is on a powerful machine. Having those commands run on modified files only would allow us to use this hook. Right now we can't.

**Comment by kenyee** [comment]
+1 looking for something like this too

**Comment by kenyee** [comment]
This is one way to do it from the other ktlint plugin: https://github.com/JLLeitschuh/ktlint-gradle/blob/d0ca26e0a04b51222ff9d3812c682d57b2142427/plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/GitHook.kt#L222


but having a top level formatKotlinFiles accepting args would be more useful IMHO.  The other way would involve running ktlint on every module.

**Comment by kenyee** [comment]
To follow up on this, I ended up using @Mahoney's xargs script above and using the raw pinterest ktlint executable instead.

If this is done via the gradle task, it takes a lot longer to run because Gradle has to start up, load the project, then run the ktlintFormatFiles task and if you have a project with 200+ modules, this is a big lag.  Use the ktlint executable provided by the pinterest ktlint repo instead and it'll be *much* faster for your users.

**Comment by Mahoney** [comment]
I wonder if some of the configuration caching in Gradle 8, plus an active gradle daemon, would invert that? But I don't have a suitably huge gradle project to test it on.

Unless doing it in docker or equivalent it's painful to have to keep an installed ktlint's version in line with this plugin - e.g. at the moment the released version of the plugin is on ktlint 0.47.1., but homebrew is on 0.48.2, and they do not agree on formatting rules.

**Comment by kenyee** [comment]
I don't think the config caching would help.
You can also tell homebrew to install a specific version.  Or you can download a specific executable from the pinterest repo releases (the latter is what I do).
